core/finality-grandpa/src/environment.rs

```diff
@@ -49,27 +52,191 @@ use crate::until_imported::UntilVoteTargetImported;
+	pub fn last(&self) -> &CompletedRound<Block> {
+		self.inner.back()
+			.expect("inner is never empty; always contains at least genesis; qed")
+	}
+
+	/// Push a new completed round, returns false if the given round is older
+	/// than the last completed round.
+	pub fn push(&mut self, completed_round: CompletedRound<Block>) -> bool {
+		if self.last().number >= completed_round.number {
+			return false;
+		}
+
+		if self.inner.len() == NUM_LAST_COMPLETED_ROUNDS {
+			self.inner.pop_front();
+		}
+
+		self.inner.push_back(completed_round);
+
+		true
+	}
+}
 
-/// A read-only view of the last completed round.
-pub(crate) struct LastCompletedRound<H, N> {
-	inner: RwLock<CompletedRound<H, N>>,
+/// The state of the current voter set, whether it is currently active or not
+/// and information related to the previously completed rounds. Current round
+/// voting status is used when restarting the voter, i.e. it will re-use the
+/// previous votes for a given round if appropriate (same round and same local
+/// key).
+#[derive(Debug, Decode, Encode, PartialEq)]
+pub enum VoterSetState<Block: BlockT> {
+	/// The voter is live, i.e. participating in rounds.
+	Live {
+		/// The previously completed rounds.
+		completed_rounds: CompletedRounds<Block>,
+		/// Vote status for the current round.
+		current_round: HasVoted<Block>,
+	},
+	/// The voter is paused, i.e. not casting or importing any votes.
+	Paused {
+		/// The previously completed rounds.
+		completed_rounds: CompletedRounds<Block>,
+	},
 }
 
-impl<H: Clone, N: Clone> LastCompletedRound<H, N> {
-	/// Create a new tracker based on some starting last-completed round.
-	pub(crate) fn new(round: CompletedRound<H, N>) -> Self {
-		LastCompletedRound { inner: RwLock::new(round) }
+impl<Block: BlockT> VoterSetState<Block> {
+	/// Returns the last completed rounds.
+	pub(crate) fn completed_rounds(&self) -> CompletedRounds<Block> {
+		match self {
+			VoterSetState::Live { completed_rounds, .. } =>
+				completed_rounds.clone(),
+			VoterSetState::Paused { completed_rounds } =>
+				completed_rounds.clone(),
+		}
+	}
+}
+
+/// Whether we've voted already during a prior run of the program.
+#[derive(Debug, Decode, Encode, PartialEq)]
+pub enum HasVoted<Block: BlockT> {
+	/// Has not voted already in this round.
+	No,
+	/// Has voted in this round.
+	Yes(AuthorityId, Vote<Block>),
+}
+
+/// The votes cast by this voter already during a prior run of the program.
+#[derive(Debug, Clone, Decode, Encode, PartialEq)]
+pub enum Vote<Block: BlockT> {
+	/// Has cast a proposal.
+	Propose(PrimaryPropose<Block>),
+	/// Has cast a prevote.
+	Prevote(Option<PrimaryPropose<Block>>, Prevote<Block>),
+	/// Has cast a precommit (implies prevote.)
+	Precommit(Option<PrimaryPropose<Block>>, Prevote<Block>, Precommit<Block>),
+}
+
+impl<Block: BlockT> HasVoted<Block> {
+	/// Returns the proposal we should vote with (if any.)
+	pub fn propose(&self) -> Option<PrimaryPropose<Block>> {
```

Comment:
any reasons why this can't return references rather than cloning? I guess they're rather light until signed